File: Leetcode/shortestDistance.java
public int shortestDistance(String[] words, String word1, String word2) {
        HashMap<String, ArrayList<Integer>> dict = new HashMap<String, ArrayList<Integer>>();
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            if (dict.containsKey(word) ) {
                dict.get(word).add(i);
            } else {
                ArrayList<Integer> indices = new ArrayList<Integer>();
                indices.add(i);
                dict.put(word, indices);
            }
        }
        ArrayList<Integer> indices1 = dict.get(word1);
        ArrayList<Integer> indices2 = dict.get(word2);
        int i =0;
        int j =0;
        int min = Integer.MAX_VALUE;
        while (i < indices1.size() && j < indices2.size()) {
            int index1 = indices1.get(i);
            int index2 = indices2.get(j);
            min = Math.min(min, Math.abs(index1 - index2));
            if (index1 < index2) {
                i++;
            } else {
                j++;
            }
        }
        return min;
    }

File: Leetcode/Python/shortestDistance.py
def shortestDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        if word1 is None or word2 is None:
            return -1
        elif word1 == word2:
            return 0
        else:
            dict = {}
            for index, word in enumerate(words):
                if word in dict:
                    dict[word].append(index)
                else:
                    dict[word] = [index]
            
            index1 = dict[word1]
            index2 = dict[word2]
            i = 0
            j = 0
            distance = len(words) + 1
            while i < len(index1) and j < len(index2):
                distance = min(distance, abs(index1[i]- index2[j]))
                if index1[i] < index2[j]:
                    i += 1
                else:
                    j += 1
            
            return distance
